File: lib/features/auth/data/auth_repository.dart
import 'package:google_sign_in/google_sign_in.dart';
import 'package:sign_in_with_apple/sign_in_with_apple.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:vitasense/core/services/cache_service.dart';

class AuthRepository {
  final SupabaseClient _client = Supabase.instance.client;

  // ─── Sign Up ─────────────────────────────────────────────────────────────────
  Future<Map<String, dynamic>> signUp({
    required String email,
    required String password,
    required String fullName,
  }) async {
    final response = await _client.functions.invoke(
      'manage-auth',
      body: {
        'action': 'sign_up',
        'email': email,
        'password': password,
        'fullName': fullName,
      },
    );
    print('SignUp response: ${response.data}');
    print('SignUp status: ${response.status}');
    final responseData = response.data as Map;
    return Map<String, dynamic>.from(responseData['data'] as Map);
  }

  // ─── Sign In ──────────────────────────────────────────────────────────────────
  Future<Map<String, dynamic>> signIn({
    required String email,
    required String password,
  }) async {
    final response = await _client.auth.signInWithPassword(
      email: email,
      password: password,
    );

    final session = response.session;
    if (session == null) {
      throw Exception('Sign in failed: no session returned');
    }

    return {
      'access_token': session.accessToken,
      'refresh_token': session.refreshToken,
      'user_id': session.user.id,
    };
  }

  // ─── Sign Out ─────────────────────────────────────────────────────────────────
  Future<void> signOut() async {
    await _client.auth.signOut();
  }

  // ─── Reset Password ───────────────────────────────────────────────────────────
  Future<void> resetPassword(String email) async {
    await _client.functions.invoke(
      'manage-auth',
      body: {
        'action': 'reset_password',
        'email': email,
      },
    );
  }

  // ─── Sign In with Google ──────────────────────────────────────────────────────
  Future<void> signInWithGoogle() async {
    final googleUser = await GoogleSignIn(
      clientId: '965374839295-0f812oks249d17kn70o2riiej0fntkek.apps.googleusercontent.com',
    ).signIn();
    if (googleUser == null) {
      throw Exception('Google sign in cancelled');
    }

    final googleAuth = await googleUser.authentication;
    final googleIdToken = googleAuth.idToken;
    if (googleIdToken == null) {
      throw Exception('Failed to retrieve Google ID token');
    }

    await _client.auth.signInWithIdToken(
      provider: OAuthProvider.google,
      idToken: googleIdToken,
    );
  }

  // ─── Sign In with Apple ───────────────────────────────────────────────────────
  Future<void> signInWithApple() async {
    final credential = await SignInWithApple.getAppleIDCredential(
      scopes: [
        AppleIDAuthorizationScopes.email,
        AppleIDAuthorizationScopes.fullName,
      ],
    );

    final appleIdToken = credential.identityToken;
    if (appleIdToken == null) {
      throw Exception('Failed to retrieve Apple identity token');
    }

    await _client.auth.signInWithIdToken(
      provider: OAuthProvider.apple,
      idToken: appleIdToken,
    );
  }

  // ─── Get User Profile ─────────────────────────────────────────────────────────
  Future<Map<String, dynamic>> getUserProfile() async {
    return CacheService().fetchWithStaleWhileRevalidate(
      key: 'user_profile',
      fetchFuture: () async {
        final userId = _client.auth.currentUser?.id;
        if (userId == null) throw Exception('Not authenticated');
        final response = await _client
            .from('profiles')
            .select('*')
            .eq('id', userId)
            .single();
        return {
          'id': response['id'],
          'email': _client.auth.currentUser?.email ?? '',
          'full_name': response['name'],
          'onboarding_completed': response['onboarding_completed'],
          'subscription_status': response['subscription_status'],
          'trial_expires_at': response['trial_expires_at'],
          'goal_type': response['goal_type'],
          'daily_calorie_target': response['daily_calorie_target'],
          'daily_protein_target': response['daily_protein_target'],
          'daily_carbs_target': response['daily_carbs_target'],
          'daily_fat_target': response['daily_fat_target'],
          'goal_pace': response['goal_pace'],
          'activity_level': response['activity_level'],
          'weight_kg': response['weight_kg'],
          'target_weight_kg': response['target_weight_kg'],
          'height_cm': response['height_cm'],
          'gender': response['gender'],
          'age': response['age'],
          'daily_water_target': response['daily_water_target'],
          'avatar_url': response['avatar_url'],
          'allergies': response['allergies'],
          'health_conditions': response['health_conditions'],
          'dietary_preferences': response['dietary_preferences'],
        };
      },
    );
  }

  // ─── Update Profile ───────────────────────────────────────────────────────────
  Future<void> updateProfile(Map<String, dynamic> data) async {
    await _client.functions.invoke(
      'manage-profile',
      body: {
        'action': 'update_profile',
        ...data,
      },
    );
    CacheService().invalidate('user_profile');
  }

  // ─── Calculate Targets ────────────────────────────────────────────────────────
  Future<Map<String, dynamic>> calculateTargets() async {
    return CacheService().fetchWithStaleWhileRevalidate(
      key: 'user_targets',
      fetchFuture: () async {
        final response = await _client.functions.invoke(
          'manage-profile',
          body: {
            'action': 'calculate_targets',
          },
        );
        final responseData = response.data as Map;
        return Map<String, dynamic>.from(responseData['data'] as Map);
      },
    );
  }

  // ─── Complete Onboarding ──────────────────────────────────────────────────────
  Future<void> completeOnboarding() async {
    await _client.functions.invoke(
      'manage-profile',
      body: {
        'action': 'complete_onboarding',
      },
    );
  }
}

File: supabase/functions/manage-auth/AuthRepository.ts
import { getSupabaseAdmin, getSupabaseAuth } from '../_shared/supabaseClient.ts';
import { ValidationError, NotFoundError } from '../_shared/errorHandler.ts';

export interface AuthResult {
  userId: string;
  email: string;
  fullName?: string;
  accessToken?: string;
  refreshToken?: string;
}

export interface UserProfile {
  id: string;
  email: string;
  fullName: string;
  onboardingCompleted: boolean;
  subscriptionStatus: string;
  goalType?: string;
  dailyCalorieTarget?: number;
  dailyProteinTarget?: number;
  dailyCarbsTarget?: number;
  dailyFatTarget?: number;
  goalPace?: string;
  activityLevel?: string;
  weightKg?: number;
  heightCm?: number;
  gender?: string;
  age?: number;
  allergies?: string[];
  healthConditions?: string[];
  dietaryPreferences?: string[];
  dailyWaterTarget?: number;
}

export class AuthRepository {
  static async signUp(
    email: string,
    password: string,
    fullName: string,
  ): Promise<AuthResult> {
    const supabaseAdmin = getSupabaseAdmin();
    const { data, error } = await supabaseAdmin.auth.signUp({
      email,
      password,
      options: {
        data: { full_name: fullName }
      }
    });
    if (error || !data.user) throw new ValidationError(error?.message ?? 'Rejestracja nie powiodła się.');

    // Auto-confirm email to allow immediate sign-in
    try {
      await supabaseAdmin.auth.admin.updateUserById(
        data.user.id,
        { email_confirm: true }
      );
    } catch (e) {
      console.warn(`[manage-auth] signUp - auto-confirm failed:`, e);
    }

    // Ensure the profile is created in the database
    try {
      const trialExpiresAt = new Date(Date.now() + 3 * 24 * 60 * 60 * 1000).toISOString();
      await supabaseAdmin
        .from('profiles')
        .upsert({
          id: data.user.id,
          name: fullName,
          subscription_status: 'trialing',
          trial_expires_at: trialExpiresAt,
        }, { onConflict: 'id' });
    } catch (e) {
      console.warn(`[manage-auth] signUp - profile upsert failed:`, e);
    }

    return {
      userId: data.user.id,
      email: data.user.email ?? email,
      fullName,
    };
  }

  static async signIn(
    email: string,
    password: string,
  ): Promise<AuthResult> {
    const supabase = getSupabaseAuth();
    const { data, error } = await supabase.auth.signInWithPassword({ email, password });
    if (error || !data.session) {
      throw new ValidationError('Nieprawidłowy email lub hasło.');
    }
    return {
      userId: data.user.id,
      email: data.user.email ?? email,
      accessToken: data.session.access_token,
      refreshToken: data.session.refresh_token,
    };
  }

  static async signOut(userId: string): Promise<void> {
    const supabase = getSupabaseAdmin();
    const { error } = await supabase.auth.admin.signOut(userId);
    if (error) throw new Error(`Błąd wylogowania: ${error.message}`);
  }

  static async resetPassword(email: string): Promise<void> {
    const admin = getSupabaseAdmin();
    const { data } = await admin.auth.admin.listUsers({ perPage: 1000 });
    const exists = (data?.users ?? []).some((u) => u.email === email);
    if (!exists) throw new NotFoundError('Nie znaleziono użytkownika z podanym adresem email.');
    const { error } = await admin.auth.resetPasswordForEmail(email, {
      redirectTo: 'vitasense://reset-password',
    });
    if (error) throw new Error(`Błąd wysyłania emaila: ${error.message}`);
  }

  static async refreshToken(refreshToken: string): Promise<AuthResult> {
    const supabase = getSupabaseAuth();
    const { data, error } = await supabase.auth.refreshSession({
      refresh_token: refreshToken,
    });
    if (error || !data.session || !data.user) {
      throw new ValidationError('Nieprawidłowy lub wygasły token odświeżania.');
    }
    return {
      userId: data.user.id,
      email: data.user.email ?? '',
      accessToken: data.session.access_token,
      refreshToken: data.session.refresh_token,
    };
  }

  static async getUser(userId: string): Promise<UserProfile> {
    const admin = getSupabaseAdmin();
    const [authRes, profileRes] = await Promise.all([
      admin.auth.admin.getUserById(userId),
      admin.from('profiles').select('*').eq('id', userId).single(),
    ]);
    if (authRes.error || !authRes.data.user) {
      throw new NotFoundError('Nie znaleziono użytkownika.');
    }
    if (profileRes.error || !profileRes.data) {
      throw new NotFoundError('Nie znaleziono profilu użytkownika.');
    }
    const u = authRes.data.user;
    const p = profileRes.data;
    return {
      id: p.id,
      email: u.email ?? '',
      fullName: p.name,
      onboardingCompleted: p.onboarding_completed === true,
      subscriptionStatus: p.subscription_status,
      goalType: p.goal_type ?? undefined,
      dailyCalorieTarget: p.daily_calorie_target ? Number(p.daily_calorie_target) : undefined,
      dailyProteinTarget: p.daily_protein_target ? Number(p.daily_protein_target) : undefined,
      dailyCarbsTarget:   p.daily_carbs_target   ? Number(p.daily_carbs_target)   : undefined,
      dailyFatTarget:     p.daily_fat_target      ? Number(p.daily_fat_target)     : undefined,
      goalPace: p.goal_pace ?? undefined,
      activityLevel: p.activity_level ?? undefined,
      weightKg: p.weight_kg ? Number(p.weight_kg) : undefined,
      heightCm: p.height_cm ? Number(p.height_cm) : undefined,
      gender: p.gender ?? undefined,
      age: p.age ? Number(p.age) : undefined,
      allergies: p.allergies ?? [],
      healthConditions: p.health_conditions ?? [],
      dietaryPreferences: p.dietary_preferences ?? [],
      dailyWaterTarget: p.daily_water_target ? Number(p.daily_water_target) : undefined,
    };
  }
}
